aws-s3-transfer-manager/src/operation/upload_objects/worker.rs

```diff
@@ -163,42 +187,56 @@ fn derive_object_key<'a>(
-                        ctx.state.successful_uploads.fetch_add(1, Ordering::SeqCst);
-
-                        ctx.state
-                            .total_bytes_transferred
-                            .fetch_add(bytes_transferred, Ordering::SeqCst);
-
-                        tracing::debug!("worker finished uploading object {key:?}");
-                    }
-                    Err(err) => {
-                        tracing::debug!("worker failed to upload object {key:?}: {err}");
-                        handle_failed_upload(err, &ctx, Some(key))?;
+    loop {
+        tokio::select! {
+            _ = cancel_rx.changed() => {
+                tracing::error!("received cancellation signal, exiting and ignoring any future work");
+                return Err(crate::error::Error::new(ErrorKind::OperationCancelled, CANCELLATION_ERROR.to_owned()));
+            }
+            job = list_directory_rx.recv() => {
+                match job {
+                    Err(_) => break,
+                    Ok(job) => {
+                        match job {
+                            Ok(job) => {
+                                let key = job.key.clone();
+                                let result = upload_single_obj(&ctx, job, cancel_rx.clone()).await;
+                                match result {
+                                    Ok(bytes_transferred) => {
+                                        ctx.state.successful_uploads.fetch_add(1, Ordering::SeqCst);
+
+                                        ctx.state
+                                            .total_bytes_transferred
+                                            .fetch_add(bytes_transferred, Ordering::SeqCst);
+
+                                        tracing::debug!("worker finished uploading object {key:?}");
+                                    }
+                                    Err(err) => {
+                                        tracing::debug!("worker failed to upload object {key:?}: {err}");
+                                        handle_failed_upload(err, &ctx, Some(key))?;
+                                    }
+                                }
+                            }
+                            Err(err) => {
+                                tracing::debug!("worker received an error from the `list_directory` task: {err}");
+                                handle_failed_upload(err, &ctx, None)?;
+                            }
+                        }
                     }
                 }
             }
-            Err(err) => {
-                tracing::debug!("worker received an error from the `list_directory` task: {err}");
-                handle_failed_upload(err, &ctx, None)?;
-            }
         }
     }
 
-    tracing::trace!("req channel closed, worker finished");
     Ok(())
 }
 
 async fn upload_single_obj(
     ctx: &UploadObjectsContext,
     job: UploadObjectJob,
+    cancel_rx: watch::Receiver<bool>,
```

Comment:
We're passing around `cancel_rx` everywhere, seems like it belongs in the context which is meant to hold the state for the operation. I realize we are storing the `tx` side of this in the handle so maybe that causes issues but if it doesn't I think it would be cleaner to just thread this into the context so that we don't have to have a separate parameter everywhere.